### PR TITLE
Ruby - Danica Reinicke - 

### PR DESCRIPTION
**Pull request by dreinicke27**

### CODE REVIEW

File: tests/test_wave_04.py [hunk 1]
```diff
@@ -44,7 +44,7 @@ def test_get_highest_word_tie_prefers_shorter_word_unsorted_list():
 
     # Act
     best_word = get_highest_word_score(words)
-
+
```
Comment (mmcknett-ada) [inline tests/test_wave_04.py:47]: [nit] Remove unnecessary additions before you submit. 😄

File: adagrams/game.py [hunk 2]
```diff
@@ -1,11 +1,151 @@
+import random
+
+POOL = {
+    "A":9,
```
Comment (mmcknett-ada) [inline adagrams/game.py:4]: [nit] space after `:`

```python
    "A": 9,
```

File: adagrams/game.py [hunk 3]
```diff
@@ -1,11 +1,151 @@
...
 def draw_letters():
-    pass
+    pool_list = []
+    chosen_letters = []
+
+    for key in POOL:
+        counter = POOL[key]
+        while counter > 0:
+            pool_list.append(key)
+            counter -= 1
```
Comment (mmcknett-ada) [inline adagrams/game.py:40]: [suggestion] Because Python is a little horrifying, you can also write this more concisely as:

```python
pool_list += counter * key
```

File: adagrams/game.py [hunk 4]
```diff
@@ -1,11 +1,151 @@
...
+
+    while len(chosen_letters) < 10:
+        letter = random.choice(pool_list)
+        index = pool_list.index(letter)
```
Comment (mmcknett-ada) [inline adagrams/game.py:44]: Built-in functions can have hidden costs. Think about how `random.choice` and `list.index` must necessarily be written, and what their runtime complexity might be. How does that compare to code that randomly selects an index, then gets the letter at that index, like this?

```python
        index = random.randint(0, len(pool_list) - 1)
        letter = pool_list[index]
```

File: adagrams/game.py [hunk 5]
```diff
@@ -1,11 +1,151 @@
...
+    
+    return chosen_letters
 
 def uses_available_letters(word, letter_bank):
-    pass
+    letters = []
+    word_cap = word.upper()
+
+    for i in range(0,len(word_cap)):
+        letters.append(word_cap[i])
```
Comment (mmcknett-ada) [inline adagrams/game.py:55]: Prefer a normal for loop for simple cases like this.

```python
    for letter in word_cap:
        letters.append(letter)
```

File: adagrams/game.py [hunk 6]
```diff
@@ -1,11 +1,151 @@
...
+            count_if_false += 1
+        elif letters.count(letter) > letter_bank.count(letter):
+            count_if_false += 1
+        else:
+            continue
+
+    if count_if_false == 0:
+        return True
+    else:
+        return False
```
Comment (mmcknett-ada) [inline adagrams/game.py:70]: Try to avoid writing code like:
```python
if some_condition:
    return True
else:
    return False
```

Instead, return the result of the condition directly, so here

```python
return count_if_false == 0
```

File: adagrams/game.py [hunk 7]
```diff
@@ -1,11 +1,151 @@
...
+    word_cap = word.upper()
+
+    for i in range(0,len(word_cap)):
+        letters.append(word_cap[i])
+
+    count_if_false = 0
+
+    for letter in letters:
+        if letter not in letter_bank:
+            count_if_false += 1
```
Comment (mmcknett-ada) [inline adagrams/game.py:61]: Since you only return `True` if `count_if_false` is 0, you can know the moment it becomes 1 that you are going to return `False`. That means you can turn both `count_if_false += 1` lines into early `return False`.

File: adagrams/game.py [hunk 8]
```diff
@@ -1,11 +1,151 @@
...
+        elif letters.count(letter) > letter_bank.count(letter):
```
Comment (mmcknett-ada) [inline adagrams/game.py:62]: [optimization] What's the big-O complexity of this loop? Could it be less? 😄

ALSO...how much does big-O complexity matter for this method? Do you think big-O is important, or would it be more important to measure the use of time/memory empirically?

File: adagrams/game.py (hunk 5 above)
Comment (mmcknett-ada) [reply adagrams/game.py:55]: Incidentally, this means `letters` is just a copy of `word_cap`. You can eliminate this loop and `word_cap` by setting `letters = word.upper()` above.

File: adagrams/game.py (hunk 8 above)
Comment (mmcknett-ada) [reply adagrams/game.py:62]: [nit] You also don't need the `if letter not in letter_bank` since you have this case.

File: adagrams/game.py [hunk 9]
```diff
@@ -1,11 +1,151 @@
...
+            count_if_false += 1
+        else:
+            continue
```
Comment (mmcknett-ada) [inline adagrams/game.py:65]: [nit] When `else: continue` is at the end of a loop, it's always redundant. You can remove it.

File: adagrams/game.py [hunk 10]
```diff
@@ -1,11 +1,151 @@
...
+        return True
+    else:
+        return False
 
 def score_word(word):
-    pass
+    score = 0
+    word = word.upper()
+    one_point_letters = ["A","E","I","O","U","L","N","R","S","T"]
+    two_point_letters = ["D","G"]
```
Comment (mmcknett-ada) [inline adagrams/game.py:76]: [suggestion] You could make each of these strings instead of lists, since `str` supports `in`

```python
two_point_letters = "DG"
```

File: adagrams/game.py [hunk 11]
```diff
@@ -1,11 +1,151 @@
...
+    four_point_letters = ["F","H","V","W","Y"]
+    five_point_letters = ["K"]
+    eight_point_letters = ["J","X"]
+    ten_point_letters = ["Q","Z"]
+
+    for letter in word:
+        if letter in one_point_letters:
+            score +=1
+        elif letter in two_point_letters:
+            score+=2
```
Comment (mmcknett-ada) [inline adagrams/game.py:87]: [nit] Try to keep your spacing consistent and always space around operators

```python
score += 1
# ...
score += 2
```

File: adagrams/game.py [hunk 12]
```diff
@@ -1,11 +1,151 @@
...
+        elif letter in three_point_letters:
+            score+=3
+        elif letter in four_point_letters:
+            score+=4
+        elif letter in five_point_letters:
+            score+=5
+        elif letter in eight_point_letters:
+            score+=8
+        else:
+            score += 10
```
Comment (mmcknett-ada) [inline adagrams/game.py:97]: [comment] Checking each letter in a range like this is definitely more maintainable than checking for each letter one by one. It could be faster, though. Consider how many checks this code does if `letter` happens to be `"Z"`. How many more than `"A"`? If performance became a concern for this code, you could consider using a dictionary to map each possible letter to a score.

Building that dictionary is a little painful (and redundant), but we could imagine starting from a dictionary that uses a string of all the letters which share a score as the key, and the score as the value. That would be very maintainable and have no redundancy. Then we could write code to convert from that structure to a dictionary where each letter is its own key. This would let us have a maintainable data representation, AND quick score lookups!

That might look like starting with your variables containing lists and the if/else chain, and adding a function to build a per-letter dictionary of scores...

```python
def build_letter_to_score():
  """ produces a dictionary of letter to score, e.g. { "A": 1, "E": 1, ... "Z": 10 } """
  ...
```

You will often find that, on the job, this kind of extra effort is worth the development time to balance performance and maintainability.

File: adagrams/game.py [hunk 13]
```diff
@@ -1,11 +1,151 @@
...
+
+    high_score = max(word_scores)
+
+    words_with_high_score = []
+    
+    
+    #remove any entries that don't have the high score 
+    for i in range(0,len(word_attributes)):
+        if word_attributes[i]["score"] == high_score:
+            words_with_high_score.append(word_attributes[i])
```
Comment (mmcknett-ada) [inline adagrams/game.py:124]: A normal for loop works here
```python
    for word_attribute in word_attributes:
        if word_attribute["score"] == high_score:
            words_with_high_score.append(word_attribute)
```

File: adagrams/game.py [hunk 14]
```diff
@@ -1,11 +1,151 @@
...
-    pass
+    score = 0
+    word = word.upper()
+    one_point_letters = ["A","E","I","O","U","L","N","R","S","T"]
+    two_point_letters = ["D","G"]
+    three_point_letters = ["B","C","M","P"]
+    four_point_letters = ["F","H","V","W","Y"]
+    five_point_letters = ["K"]
+    eight_point_letters = ["J","X"]
+    ten_point_letters = ["Q","Z"]
```
Comment (mmcknett-ada) [inline adagrams/game.py:81]: [nit] Technically, you don't use `ten_point_letters`.

File: adagrams/game.py [hunk 15]
```diff
@@ -1,11 +1,151 @@
...
+    word_scores = []
+    winning_word = ""
+    winning_score = 0
+    
+    for word in word_list:
+        word_attributes.append({"word":word,"score":score_word(word),"length":len(word)})
+        word_scores.append(score_word(word))
+
+    
+
```
Comment (mmcknett-ada) [inline adagrams/game.py:115]: [nit] Avoid double- and triple-spacing in functions

File: adagrams/game.py [hunk 16]
```diff
@@ -1,11 +1,151 @@
...
+    if len(words_with_high_score) == 1:
+        winning_word = words_with_high_score[0]["word"]
+        winning_score = words_with_high_score[0]["score"]
+    #tie 
+    elif len(words_with_high_score) > 1:
+        word_lengths = []
+        for i in range(0,len(words_with_high_score)):
+            word_lengths.append(words_with_high_score[i]["length"])
+        
+        max_length = max(word_lengths)
```
Comment (mmcknett-ada) [inline adagrams/game.py:138]: [suggestion] In my experience, matching indices between lists is risky because the lists can get out of sync as code changes. This implementation of `get_highest_word_score` is a little more complex than it could be.

Sometimes when we do these kinds of algorithms that have confusing layers of logical requirements, it's easy to get stuck in a corner by one early-on decision. In this case, I think you got cornered by trying to do this function with the built-in `max` and `min` functions. You'll probably find that if you start rewriting this function from scratch and begin with calculating "max" using a traditional for loop, you'll end up with a simpler, but still correct, result.

**Review by mmcknett-ada — COMMENTED** [review]
Looks great! 🟢 Green!

I have some nitpicks, comments, and suggestions, but those don't detract from the fact that this is a solid submission. Very nice!

Some overall feedback:
- The tests pass and your code even satisfies the corner cases I've found while grading that the tests don't cover.
- Your `draw_letters` function is fully correct, which is rare. Great job!
- You have lots of opportunities to simplify redundant statements. Consider doing a pass to refactor and clean up your submissions once you've got all the tests passing.
